apps/glusterfs/operations.go

```diff
@@ -457,6 +457,152 @@ func (vdel *VolumeDeleteOperation) Finalize() error {
 	})
 }
 
+// VolumeCloneOperation implements the operation functions used to
+// clone an existing volume.
+type VolumeCloneOperation struct {
+	OperationManager
+	noRetriesOperation
+
+	// The volume to use as source for the clone
+	vol *VolumeEntry
+	// Optional name for the new volume
+	clonename string
+	// The newly cloned volume, will be set in Exec()
+	clone *VolumeEntry
+	// The bricks for the clone
+	bricks []*BrickEntry
+	// The devices of the bricks
+	devices []*DeviceEntry
+}
+
+func NewVolumeCloneOperation(
+	vol *VolumeEntry, db wdb.DB, clonename string) *VolumeCloneOperation {
+
+	return &VolumeCloneOperation{
+		OperationManager: OperationManager{
+			db: db,
+			op: NewPendingOperationEntry(NEW_ID),
+		},
+		vol:       vol,
+		clonename: clonename,
+		clone:     nil,
+	}
+}
+
+func (vc *VolumeCloneOperation) Label() string {
+	return "Create Clone of a Volume"
+}
+
+func (vc *VolumeCloneOperation) ResourceUrl() string {
+	return fmt.Sprintf("/volumes/%v", vc.clone.Info.Id)
+}
+
+func (vc *VolumeCloneOperation) Build() error {
+	return vc.db.Update(func(tx *bolt.Tx) error {
+		vc.op.RecordCloneVolume(vc.vol)
+		clone, bricks, devices, err := vc.vol.prepareVolumeClone(tx, vc.clonename)
+		if err != nil {
+			return err
+		}
+		vc.clone = clone
+		vc.bricks = bricks
+		vc.devices = devices
+		vc.op.RecordAddVolumeClone(vc.clone)
+		// record new bricks
+		for _, b := range bricks {
+			vc.op.RecordAddBrick(b)
+			if e := b.Save(tx); e != nil {
+				return e
+			}
+		}
+		// save device updates
+		for _, d := range vc.devices {
+			if e := d.Save(tx); e != nil {
+				return e
+			}
+		}
+		// record changes to parent volume
+		if e := vc.vol.Save(tx); e != nil {
+			return e
+		}
+		// add the new volume to the cluster
+		c, err := NewClusterEntryFromId(tx, vc.clone.Info.Cluster)
+		if err != nil {
+			return err
+		}
+		c.VolumeAdd(vc.clone.Info.Id)
+		if err := c.Save(tx); err != nil {
+			return err
+		}
+		// save the pending operation
+		if e := vc.op.Save(tx); e != nil {
+			return e
+		}
+		return nil
+	})
+}
+
+func (vc *VolumeCloneOperation) Exec(executor executors.Executor) error {
+	vcr, host, err := vc.vol.cloneVolumeRequest(vc.db, vc.clone.Info.Name)
+	if err != nil {
+		return err
+	}
+
+	// get all details of the original volume (order of bricks etc)
+	orig, err := executor.VolumeInfo(host, vc.vol.Info.Name)
+	if err != nil {
+		return err
+	}
+
+	clone, err := executor.VolumeClone(host, vcr)
+	if err != nil {
+		return err
+	}
+
+	if err := updateCloneBrickPaths(vc.bricks, orig, clone); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (vc *VolumeCloneOperation) Rollback(executor executors.Executor) error {
+	return vc.db.Update(func(tx *bolt.Tx) error {
```

Comment:
Looks like some stuff is missing from rollback. I think we need to go to the executor and clean up any snapshots/bricks/etc that the operation may have created before the server was terminated. If you don't think you can fit it in thss PR I'd be OK with that, but we should put in a TODO comment at the very least.

Reply:
For now a TODO comment is all I can do. Once the more advanced Snatshot operations are available, the VolumeClone operation can move from the CmdExecutor to `operations.go` where each step can be rolled back.

Reply:
That's fine. Please add the comment when you have a moment. Thanks.